### PR TITLE
Feature/liquidstaking2.0

### PR DESCRIPTION
**Pull request by alannotnerd**

### CODE REVIEW

**Comment by alannotnerd** [comment]
```
2021-09-01T04:33:06.509Z | debug: "Current rebonding amount is : 60000000000000"
2021-09-01T04:33:06.509Z | debug: "task::inserting <-> [liquidStaking, settlement, 60000000000000]"
2021-09-01T04:33:06.658Z | debug: "task::handling <-> [liquidStaking, settlement, 60000000000000]"
2021-09-01T04:33:06.683Z | debug: "tx::processing"
2021-09-01T04:33:18.572Z | debug: "Para: liquidStaking::StakingOpRequest <-> [0,0,0]"
2021-09-01T04:33:30.063Z | debug: "tx::succeeded"
2021-09-01T04:33:30.064Z | debug: "task::removing <-> [g1ybxss2te0wvRN6emzLdEXjbVzbKDuA]"
2021-09-01T04:34:06.508Z | debug: "Current rebonding amount is : 60000000000000"
2021-09-01T04:34:06.509Z | debug: "task::inserting <-> [liquidStaking, settlement, 60000000000000]"
2021-09-01T04:34:06.690Z | debug: "task::handling <-> [liquidStaking, settlement, 60000000000000]"
2021-09-01T04:34:06.714Z | debug: "tx::processing"
2021-09-01T04:34:18.048Z | debug: "Para: liquidStaking::StakingOpRequest <-> [0,0,0]"
2021-09-01T04:34:30.056Z | debug: "tx::succeeded"
2021-09-01T04:34:30.056Z | debug: "task::removing <-> [g1ybysi2u2Z3UqfL2ZQ6hwArScZSdqBt]"
2021-09-01T04:34:30.582Z | debug: "Para: liquidStaking::Staked <-> [\"hJKzPoi3MQnSLvbShxeDmzbtHncrMXe5zwS3Wa36P6kXeNpcv\",5000000000000]"
2021-09-01T04:35:06.509Z | debug: "Current rebonding amount is : 60000000000000"
2021-09-01T04:35:06.510Z | debug: "task::inserting <-> [liquidStaking, settlement, 60000000000000]"
2021-09-01T04:35:06.683Z | debug: "task::handling <-> [liquidStaking, settlement, 60000000000000]"
2021-09-01T04:35:06.706Z | debug: "tx::processing"
2021-09-01T04:35:18.150Z | debug: "Para: liquidStaking::StakingOpRequest <-> [0,5000000000000,0]"
2021-09-01T04:35:18.151Z | debug: "task::inserting <-> [staking, rebond, 5000000000000]"
2021-09-01T04:35:18.224Z | debug: "Para: liquidStaking::StakingOpRequest <-> [0,5000000000000,0]"
2021-09-01T04:35:18.226Z | debug: "task::inserting <-> [staking, rebond, 5000000000000]"
2021-09-01T04:35:30.060Z | debug: "tx::succeeded"
2021-09-01T04:35:30.061Z | debug: "task::removing <-> [g1ybzs83rtsIOD7sRHBVhgV47bw22xft]"
2021-09-01T04:35:30.562Z | debug: "task::handling <-> [staking, rebond, 5000000000000]"
2021-09-01T04:35:30.625Z | debug: "tx::processing"
2021-09-01T04:35:36.035Z | debug: "Relay: staking::Bonded <-> [\"5DAAnrj7VHTznn2AWBemMuyBwZWs6FNFjdyVXUeYum3PTXFy\",5000000000000]"
2021-09-01T04:35:36.061Z | debug: "tx::succeeded"
2021-09-01T04:35:36.061Z | debug: "task::removing <-> [g1ybzz5m8eoE1nIPLLfh43AnCw2khyOC]"
2021-09-01T04:35:36.564Z | debug: "task::handling <-> [staking, rebond, 5000000000000]"
2021-09-01T04:35:36.591Z | debug: "tx::processing"
2021-09-01T04:35:42.024Z | debug: "Relay: staking::Bonded <-> [\"5DAAnrj7VHTznn2AWBemMuyBwZWs6FNFjdyVXUeYum3PTXFy\",5000000000000]"
2021-09-01T04:35:42.048Z | debug: "tx::succeeded"
2021-09-01T04:35:42.049Z | debug: "task::removing <-> [g1ybzz77wqWxbTylqLYAqRV0Sdp7vpBj]"
2021-09-01T04:36:06.509Z | debug: "Current rebonding amount is : 50000000000000"
2021-09-01T04:36:06.510Z | debug: "task::inserting <-> [liquidStaking, settlement, 50000000000000]"
2021-09-01T04:36:06.633Z | debug: "task::handling <-> [liquidStaking, settlement, 50000000000000]"
2021-09-01T04:36:06.655Z | debug: "tx::processing"
2021-09-01T04:36:18.598Z | debug: "Para: liquidStaking::StakingOpRequest <-> [0,0,0]"
2021-09-01T04:36:30.063Z | debug: "tx::succeeded"
2021-09-01T04:36:30.063Z | debug: "task::removing <-> [g1yc0ry4e2cKAeZRDPIWcqSoMHTlrGHm]"
```

**Comment by alannotnerd** [comment]
Seems all feature work.

File: src/main.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ program
   .option('-h, --redis-host <string>', 'Redis server host', 'localhost')
   .option('-p, --redis-port <number>', 'Redis server port', '6379')
   .option('-s, --seed <string>', 'The account seed to use', '//Dave')
+  .option('-c, --controller <string>', 'The controller seed to use', '//Eve')
```
Comment (GopherJ) [inline src/main.ts:27]: we should use another test account other than `Eve` because `Eve` is now used by nominate client

File: src/service.ts [hunk 2]
```diff
@@ -59,12 +67,14 @@ export class ApiService extends Queue {
...
     redisHost,
+    controller,
     redisPort
   }: ApiServiceConfig) {
     super({ host: redisHost, port: redisPort })
 
     this.relayWs = relayWs
     this.paraWs = paraWs
+    this.controller = controller
     this.signer = signer
```
Comment (GopherJ) [inline src/service.ts:78]: signer should be renamed because now you have `stash` and `controller`

File: src/main.ts (hunk 1 above)
Comment (alannotnerd) [reply src/main.ts:27]: Which one you suggest?

File: src/service.ts [hunk 3]
```diff
@@ -106,12 +116,26 @@ export class ApiService extends Queue {
   }
 
   public async connect(): Promise<void> {
-    this.relayApi = new ApiPromise({
-      provider: new WsProvider(this.relayWs)
-    })
+    this.relayApi = new ApiPromise(
+      options({
+        provider: new WsProvider(this.relayWs)
```
Comment (GopherJ) [inline src/service.ts:121]: we don't need `options` for relaychain I think, because it wraps our types

File: src/service.ts [hunk 4]
```diff
@@ -106,12 +116,26 @@ export class ApiService extends Queue {
...
+      })
+    )
 
     this.paraApi = new ApiPromise(
-      options({ provider: new WsProvider(this.paraWs) })
+      options({
+        provider: new WsProvider(this.paraWs),
+        types: {
```
Comment (GopherJ) [inline src/service.ts:128]: can you add missing types to parallelfinance/parallel -> resouces/types.json and parallelfinance/parallel-js? maybe @yz89 can help you

we can do this when it's stable and we are sure things wont change anymore, now it's ok to keep

File: src/service.ts [hunk 5]
```diff
@@ -166,11 +173,33 @@ export class ApiService extends Queue {
...
     })
 
+    await this.subscribe(SubscribeType.Rebonding, {
+      fn: this.relayApi.query.staking.ledger,
+      params: [this.controller.address],
+      transform: (ledger: Option<StakingLedger>) =>
+        ledger
+          .unwrap()
+          .unlocking.map((l) => l.value.toNumber())
+          .reduce((cur, total) => cur + total, 0)
```
Comment (GopherJ) [inline src/service.ts:183]: you can maybe use `sum` from `lodash` to make it looks cleaner

File: src/service.ts [hunk 6]
```diff
@@ -166,11 +173,33 @@ export class ApiService extends Queue {
...
+      const rebonding = this.storageMap.get(SubscribeType.Rebonding)
+      logger.debug(`Current rebonding amount is : ${rebonding}`)
+      this.pushBack({
+        type: ApiType.PARA,
+        pallet: ParaPalletType.LiquidStaking,
+        call: ParaCallType.Settlement,
+        params: [rebonding],
+        signer: 'stash'
+      })
+    }, 1 * 60 * 1000)
```
Comment (GopherJ) [inline src/service.ts:201]: this interval should be a constant so that we can change it

File: src/service.ts [hunk 7]
```diff
@@ -272,29 +279,38 @@ export class ApiService extends Queue {
...
-            params: !bonded
-              ? [this.signer.address, amount.toJSON(), 'Staked']
-              : [amount.toJSON()]
-          })
-
-          const validators = this.storageMap.get(SubscribeType.ValidatorList)
-          if (validators?.length) {
+        logger.debug(`Para: ${name} <-> ${event.data.toString()}`)
+        if (ParaStakingEvent.StakingOpRequest == name) {
+          const bond_amount = event.data[0].toString()
```
Comment (GopherJ) [inline src/service.ts:284]: you can use `map` from lodash to to this conversion together

File: src/service.ts [hunk 8]
```diff
@@ -272,29 +279,38 @@ export class ApiService extends Queue {
...
+          const rebond_amount = event.data[1].toString()
+          const unbond_amount = event.data[2].toString()
+          if (parseInt(bond_amount) > 0) {
```
Comment (GopherJ) [inline src/service.ts:287]: paseInt should be in previous map conversion

File: src/service.ts (hunk 7 above)
Comment (GopherJ) [inline src/service.ts:284]: also can we define a struct in backend instead of using `0` `1` 2`

File: src/service.ts (hunk 6 above)
Comment (alannotnerd) [reply src/service.ts:201]: it wiil be removed then.

File: src/service.ts (hunk 7 above)
Comment (alannotnerd) [reply src/service.ts:284]: event data is always converted to a tuple.

File: src/service.ts (hunk 4 above)
Comment (yz89) [reply src/service.ts:128]: updated metadata on parallel-js. https://github.com/parallel-finance/parallel-js/pull/3

**Review by GopherJ — APPROVED** [review]

**Pull request merged**